ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java

```diff
@@ -744,16 +746,34 @@ private void onExecutionPayloadResult(
 
   private void updateForkChoiceForImportedBlock(
       final SignedBeaconBlock block,
+      final boolean shouldApplyProposerBoost,
       final BlockImportResult result,
       final ForkChoiceStrategy forkChoiceStrategy) {
 
+    final ChainHead currentHead = recentChainData.getChainHead().orElseThrow();
+
     final SlotAndBlockRoot bestHeadBlock = findNewChainHead(forkChoiceStrategy);
-    if (!bestHeadBlock.getBlockRoot().equals(recentChainData.getBestBlockRoot().orElseThrow())) {
+    if (!bestHeadBlock.getBlockRoot().equals(currentHead.getRoot())) {
       recentChainData.updateHead(bestHeadBlock.getBlockRoot(), bestHeadBlock.getSlot());
       if (bestHeadBlock.getBlockRoot().equals(block.getRoot())) {
         result.markAsCanonical();
       }
     }
+
+    if (!result.isBlockOnCanonicalChain() && shouldApplyProposerBoost) {
+      // This is likely a reorging block that requires a full processHead to update the head.
+      // Running processHead here will ensure:
+      //
+      // - node producing a reorging block will attest and produce sync committee for the reorging
+      // block
+      // - node just importing (not producing it) a reorging block will produce sync committee for
+      // the new head
+      //
+      // It wouldn't be enough to trigger processHead at attestationDue because it will only
+      // cover attestation generation and not sync committee message, since it is generated
+      // VC side and requires head event to be sent before attestationDue.
+      processHead().finish(error -> LOG.error("Fork choice updating head failed", error));
```

Comment:
what does the user do when they see that error several times in a minute? its a very generic message, i could only see them raising a bug...

Reply:
well the exception trace will possibly cause more problems than it solves, at least we only want this infrequently so rate limiting it if we're including the whole exception trace. otherewise potentially busting it to a lower level and only including the message by default rather than the stacktrace would at least make it less noise